python-package/lightgbm/basic.py

```diff
@@ -1444,17 +1462,21 @@ def get_data(self):
         """
         if self.handle is None:
             raise Exception("Cannot get data before construct Dataset")
-        if self.data is not None and self.used_indices is not None and self.need_slice:
-            if isinstance(self.data, np.ndarray) or scipy.sparse.issparse(self.data):
-                self.data = self.data[self.used_indices, :]
-            elif isinstance(self.data, DataFrame):
-                self.data = self.data.iloc[self.used_indices].copy()
-            elif isinstance(self.data, DataTable):
-                self.data = self.data[self.used_indices, :]
-            else:
-                warnings.warn("Cannot subset {} type of raw data.\n"
-                              "Returning original raw data".format(type(self.data).__name__))
+        if self.need_slice and self.used_indices is not None and self.reference is not None:
+            self.data = self.reference.data
+            if self.data is not None:
+                if isinstance(self.data, np.ndarray) or scipy.sparse.issparse(self.data):
+                    self.data = self.data[self.used_indices, :]
+                elif isinstance(self.data, DataFrame):
+                    self.data = self.data.iloc[self.used_indices].copy()
+                elif isinstance(self.data, DataTable):
+                    self.data = self.data[self.used_indices, :]
+                else:
+                    warnings.warn("Cannot subset {} type of raw data.\n"
+                                  "Returning original raw data".format(type(self.data).__name__))
             self.need_slice = False
+        if self.data is None:
+            warnings.warn("The raw data is released, please set `free_raw_data=False`")
```

Comment:
@guolinke Sure, the warning is needed, and it's cool that you managed to avoid it when `get_data` isn't called explicitly.

Maybe it's good to uniform it with other similar warnings?

https://github.com/microsoft/LightGBM/blob/ebc831bc53b746c5cc079955364008f591877015/python-package/lightgbm/basic.py#L1229-L1230
https://github.com/microsoft/LightGBM/blob/ebc831bc53b746c5cc079955364008f591877015/python-package/lightgbm/basic.py#L1244-L1245
https://github.com/microsoft/LightGBM/blob/ebc831bc53b746c5cc079955364008f591877015/python-package/lightgbm/basic.py#L1270-L1271